Context Free Grammar Constrained Decoding (ebnf interface, compatible with llama-cpp)

Saibo-creator opened the pull request:
# What does this PR do?

This PR adds a new feature (Context Free Grammar Constrained Decoding) to the library. 
There is already one PR(WIP) for this feature( #26520 ), but this one has a different motivation and implementation.

This implementation is inspired by and adapted from  https://github.com/Shopify/torch-grammar and https://github.com/ggerganov/llama.cpp/pull/1773/files

This implementation aims to achieve the following goals:
- CFG-constrained decoding 
- EBNF notation as interface for the grammar
- standalone implementation of the grammar parser(left recursive-descent parsing, the same as in llama-cpp)
- compatibility with grammars in the llama.cpp library(https://github.com/ggerganov/llama.cpp/tree/master/grammars)
- incremental parsing and also non-incremental parsing(some tokenizers doesn't support incremental parsing from my experiments)
- unicode support for the grammar(not trivial but important for any multi-lingual model)

The two main differences from PR #26520 :
- dependency on lark, which may not be a bad thing, but my experience is that it will reduce the flexibility and may be hard to adapt to our specfic need, e.g. unicode grammar support.
- ebnf interface. This PR supports the same EBNF as in llama-cpp, so that users can directly migrate from llama-cpp

Challenges for this PR:
- compatibility with all the tokenizers in the transformers library.


Current status:
- [x] The grammar parser is implemented and works well with the example grammars from llama.cpp library.
- [x] A few integration tests are added to test the combination of grammar and tokenizer.
- [x] no unicode support yet, means it will probably fail when you want to constrain with emoji or other unicode characters. 
- [x] greedy search
- [x] sampling, top-k, top-p
- [ ] beam search  

TODO:
- [x]  Batching support
- [x] compatible with greedy decoding and sampling under `beam=1` 
- [x] ~~grammar parser fails to parse llama-cpp's json grammar(more precisely the string line). Currently, a slightly simplified version of json grammar if used~~(now fixed)
- [x] ~~grammar parser requires the last rule ending with a new line, otherwise, parsing error will be raised. This is not user-friendly and should be fixed~~ 
- [ ]  The EOS token seems not always included in the allowed tokens even when it should be, maybe due to the nature of recursive-descent parsing ?
- [ ] compatible with `beam_search` and `beam_sample`(Now throws error `RuntimeError: probability tensor contains either `inf`, `nan` or element < 0`). A good reference is the `ConstrainedBeamSearchScorer`
- [ ] unicode support
- [ ] properly test with different tokenizers(bpe, wordpiece, unigram, etc.)


<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # https://github.com/huggingface/transformers/issues/25778
Related to PR # https://github.com/huggingface/transformers/pull/26520


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

@gante 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @pacman100

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

Saibo-creator committed:
Draft Context Free Grammar Constrained Decoding

jvhoffbauer commented:
I think it is a great idea to be compatible with llama.cpp!

abhinavkulkarni commented:
Hey @Saibo-creator,

I tried running `grammar_utils.py` on `json.gbnf`, but I get the following error:

```bash
Traceback (most recent call last):
  File "/home/user/grammar.py", line 657, in <module>
    state = parse_ebnf(input_text)
  File "/home/user/grammar.py", line 249, in parse_ebnf
    grammar_repr = parse_rule(state, grammar_repr)
  File "/home/user/grammar.py", line 231, in parse_rule
    pos = parse_alternates(state, pos, name, rule_id, False)
  File "/home/user/grammar.py", line 212, in parse_alternates
    while pos[0] == "|":
IndexError: string index out of range
```

Saibo-creator commented:
Hello @abhinavkulkarni ,
     This is probably because the missing new line at the end of the grammar. 
     
Try 
```
root   ::= object

object ::= "{" ws ( string ":" ws value ("," ws string ":" ws value)* )? "}" ws

value  ::= object | array | string | number | ("true" | "false" | "null") ws

array  ::= "[" ws ( value ("," ws value)* )? "]" ws

string ::= "\"" ( [a-zA-Z0-9] )* "\"" ws

number ::= ("-"? ([0-9] | [1-9] [0-9]*)) ("." [0-9]+)? ([eE] [-+]? [0-9]+)? ws


ws ::= ([ \t\n] ws)?

```

instead of 

```
root   ::= object

object ::= "{" ws ( string ":" ws value ("," ws string ":" ws value)* )? "}" ws

value  ::= object | array | string | number | ("true" | "false" | "null") ws

array  ::= "[" ws ( value ("," ws value)* )? "]" ws

string ::= "\"" ( [a-zA-Z0-9] )* "\"" ws

number ::= ("-"? ([0-9] | [1-9] [0-9]*)) ("." [0-9]+)? ([eE] [-+]? [0-9]+)? ws


ws ::= ([ \t\n] ws)?
```

Let me know if this doesn't work

Saibo-creator committed:
add support for batch decoding

Saibo-creator committed:
integrate GCD to `generate` function api

Saibo-creator commented:
@abhinavkulkarni 

Thank you for testing!

I will look into this issue! 

By the way, I just integrated the `gcd` feature into `generation_api`, now you can run it with

```python
from transformers import AutoModelForCausalLM, AutoTokenizer
from transformers.generation.grammar_utils import IncrementalGrammarConstraint


if __name__ == '__main__':
    torch.manual_seed(2)

    model_id = "gpt2"
    tokenizer = AutoTokenizer.from_pretrained(model_id)
    tokenizer.pad_token = tokenizer.eos_token
    model = AutoModelForCausalLM.from_pretrained(model_id)
    with open("examples/grammars/json.gbnf", "r") as file:
        grammar_str = file.read()
    grammar = IncrementalGrammarConstraint(grammar_str, "root", tokenizer)

    prefix1= "This is a valid json string for email:"
    prefix2= "This is a valid json string for shopping cart:"
    input_ids = tokenizer([prefix1, prefix2],add_special_tokens=False, return_tensors="pt", padding=True)["input_ids"]

    output = model.generate(input_ids, do_sample=False, max_length=30, num_beams=2, grammar=grammar,
                            num_return_sequences=2)
    # decode output
    generations = tokenizer.batch_decode(output, skip_special_tokens=True)
    print(generations)

    """
    'This is a valid json string for email:{ "title": "Theory", "text": "Theory", "type": "text", "text": "Theory", "type',
    'This is a valid json string for shopping cart:{ "name": "MyCart", "price": "10", "price": "10", "price": "10", "price": "'
    """
```

If you have time, could you try to call via above api and confirm if the problem remains. 

For GPT2, it works as expected

abhinavkulkarni commented:
For prompt:

```
prompt = f"A sample JSON for employee record in a database: "
```

I do get a JSON-looking response, but then again, the model continues to output newlines until it hits the token limit:

```
{
   "id": 1,
   "name": "John",
   "age": 25,
   "salary": 30000,
   "department": {
       "id": 1,
       "name": "Sales"
   }
empty stack
}
....
....
....
```

Saibo-creator committed:
fix: add stop condition that if stacks is empty, the eos_token should be accepted

Saibo-creator committed:
remove trailing white space from json grammar

Saibo-creator commented:
@abhinavkulkarni 
I'm able to reproduce the "strange behavior" you reported and it actually not a bug but rather an "expected behavior".

In the json grammar, we have 
`object ::= "{" ws ( string ":" ws value ("," ws string ":" ws value)* )? "}" ws`

and the last `ws` basically allows the model to generate arbitrary white space(including new line) after the json object because such white space doesn't break the json syntax.

This may not be a desired behavior, so I removed that `ws` from the json grammar and it seems to work correctly.

But it does surprise me that the model didn't pick `EOS` after finishing the json object. Maybe the EOS was not added to the allowed token list due to a bug.
I will double-check if I treated the EOS correctly in the grammar implementation.

abhinavkulkarni commented:
@Saibo-creator: Thanks for the changes. 

Removing the whitespace `ws` fixes the newline problem.

For the simple prompt, `prompt = f"A sample JSON for employee record in a database: "`, I still see a `WARN` log line:

```
WARNING:grammar_utils:empty stack
```

A few points:

1. Should the grammar processor not reset its state after one call of `model.generate`? Calling `model.generate` on the same grammar processor throws an expection.
2. Should amping up the `repetition_penalty` not fix the whitespace issue?

Saibo-creator committed:
downgrade "empty stack" from warning to debug.

Saibo-creator commented:
@abhinavkulkarni 

Regarding resetting the state of grammar processor, here is my consideration:

Currently the `GrammarConstrainedLogitsProcessor` contains the parsing state, and I think it may be useful to not reset the state after every generation, because this could allow the user to continue the grammar-constrained generation, see the code example below.

And if the user wants to start a new generation, a new instance of `LogitProcessor` is indeed needed (here we can also add a `reset` method to make it more user-friendly)

>  It would be expensive to have to parse the grammar afresh for every single model.generate call.

I don't get this point though. 
- If the user's goal is to start another generation from scratch, then the grammar has to be parsed afresh, I don't think there is a way to avoid it ?
- If the user's goal is to continue the generation, then the example I showed below should solve the problem. The parsing would not need to start from scratch but simply continue from the old parsing state

Does this sound reasonable for you?

Regarding the design choice to put the parsing state inside the `LogitProcessor`, I'm not sure if this is the best way to do it. So I would like to have your opinion @gante  :)





```python
from transformers import AutoModelForCausalLM, AutoTokenizer,TextStreamer, set_seed
from transformers.generation.grammar_utils import IncrementalGrammarConstraint
from transformers.generation.logits_process import GrammarConstrainedLogitsProcessor


if __name__ == '__main__':

    import logging
    logging.getLogger("transformers.generation").setLevel(logging.INFO)

    # model_id = "saibo/llama-1B"
    model_id = "gpt2"
    tokenizer = AutoTokenizer.from_pretrained(model_id)
    streamer = TextStreamer(tokenizer, skip_special_tokens=True)
    tokenizer.pad_token = tokenizer.eos_token
    model = AutoModelForCausalLM.from_pretrained(model_id)
    with open("examples/grammars/json.gbnf", "r") as file:
        grammar_str = file.read()
    grammar = IncrementalGrammarConstraint(grammar_str, "root", tokenizer)

    prefix1= "This is a valid json string for email:"
    prefix2= "This is a valid json string for shopping cart:"
    input_ids = tokenizer([prefix2],add_special_tokens=False, return_tensors="pt", padding=True)["input_ids"]

    logits_processor = GrammarConstrainedLogitsProcessor(grammar)

    ###################################################
    # generation under the Grammar constraint for 10 tokens
    ##################################################

    output = model.generate(input_ids, do_sample=False, max_new_tokens=10, num_beams=2, logits_processor=[logits_processor],
                            num_return_sequences=1, repetition_penalty=1.5)

    generations = tokenizer.batch_decode(output, skip_special_tokens=True)
    print(generations)
    # 'This is a valid json string for shopping cart:{ "name": "MyCart", "price'

    ###################################################
    # Continue the generation under the same constraint for 10 tokens
    #
    # 1. Need to use the output of the previous generation as the input for the next generation
    # 2. Reuse the same logits_processor because the parser state is stored in the logits_processor
    #
    ##################################################

    output = model.generate(output[0].unsqueeze(0), do_sample=False, max_new_tokens=10, num_beams=2, logits_processor=[logits_processor],
                            num_return_sequences=1, repetition_penalty=1.5)
    generations = tokenizer.batch_decode(output, skip_special_tokens=True)
    print(generations)
    # 'This is a valid json string for shopping cart:{ "name": "MyCart", "price": "10", "description": "MyCart'


    ###################################################
    # We want to generate another valid json string
    #
    # 1. Create a new logits_processor with empty parser state
    # 2. Use the same prompt as the input
    ##################################################

    logits_processor = GrammarConstrainedLogitsProcessor(grammar)

    output = model.generate(input_ids, do_sample=True, max_new_tokens=20, num_beams=2, logits_processor=[logits_processor],
                            num_return_sequences=1, repetition_penalty=1.5)

    generations = tokenizer.batch_decode(output, skip_special_tokens=True)
    print(generations)
    # 'This is a valid json string for shopping cart:{ "name": "MyCart", "price": "10", "description": "MyCart'

```

Saibo-creator committed:
Fix: Raise Error when the input tokens are not consistent with the grammar parsing state

abhinavkulkarni commented:
> Does this sound reasonable for you?

Thanks @Saibo-creator, it makes sense not to reset the grammar state so that the user can continue the generation.

One more minor correction, the rule for generating `string` in JSON grammar should be:

```
string ::= "\"" ( [ a-zA-Z0-9] )* "\"" was
```

instead of 

```
string ::= "\"" ( [a-zA-Z0-9] )* "\"" was
```

arshadshk reviewed (approved):
Looks good!

arshadshk commented:
can we add a python gbnf file too ? Can take inspiration from : https://github.com/ggerganov/llama.cpp/blob/master/grammars/c.gbnf

Saibo-creator committed:
fix: fix parsing error with \\ by adding this case as escape char

Saibo-creator committed:
use the same json grammar as in llama-cpp

Saibo-creator committed:
feat: add c grammar

Saibo-creator committed:
hotfix: rolling json.gbnf back to the previous version. Parsing is not working as expected

Saibo-creator committed:
fix newline error, now the same state as in commit 3e78f00 in llama-cpp(Jun 12)

shermansiu commented:
How does this differ from the constrained decoding in [Picard][https://arxiv.org/abs/2109.05093] and [Synchromesh](https://openreview.net/forum?id=KmtVD97J43e)?

Saibo-creator committed:
add tests for grammar-constrained decoding

Saibo-creator commented:
Hello @gante,

I'm excited to share that the feature is now in great shape, and I'm eager to hear your thoughts on it.

The implementation of the grammar-constrained decoding feature is quite complex, as we aim to make it compatible with beam search, sampling, and all tokenizers. It's relatively straightforward to integrate it with greedy search or greedy sampling. This leads me to my first question: Should we break down this feature into multiple versions, starting with a basic one, or would it be better to aim for a comprehensive solution in a single merge? From my perspective, once we have thoroughly tested greedy decoding and greedy sampling, it might be beneficial to merge them first, as they already cater to a wide range of use cases.

Additionally, I'm facing some challenges in devising tests for this feature. Currently, I have a setup similar to what's outlined [here](https://github.com/Saibo-creator/transformers-gcd/blob/feature/cfgcd/test_grammar_constrained_decoding.py), where I create simple grammars and verify the accuracy of the generation. However, establishing a systematic testing approach is tricky. For example, if we want to test the json grammar compatibility with all models, running the model with actual weights becomes necessary. Without the weights, the model might generate nonsensical but syntactically correct json outputs, which doesn't help in effective testing. While using actual weights does lead to valid json generation, it significantly slows down the process.

I'd appreciate your insights on how to navigate these testing challenges. In the meantime, I'll continue refining the feature.

Saibo-creator committed:
1.run code cleaning
2.remove `grammar` from generation argument list, use `GrammarLogitsProcessor` instead

Saibo-creator committed:
support top-k and top-p sampling

arshadshk commented:
@ Saibo-creator I suggest we break down this feature into multiple versions, starting with a basic one. This creates motivation and encourages more people to collaborate, a greedy search for JSON sounds good for a start.

Saibo-creator commented:
Thank you @arshadshk for the feedback. I agree with you. Now I'm working on crafting tests. It's a bit challenging to write tests for this feature and I will start with tests like [this](https://github.com/huggingface/transformers/pull/27557/files#diff-368e1430d055143f6037ab2eb2eee4c15840b5d79e1cc68630313c75e57036f6)

Saibo-creator commented:
btw, @arshadshk, if you have time, could you also have a look at #27676 ? That PR tries to fix a bug which is important for this CFG feature to work properly, Thanks !

arshadshk commented:
@Saibo-creator  the (https://github.com/huggingface/transformers/issues/27676) fix makes sense, I wonder if we open up probs for <eos> too along with <pad>, we might need to terminate the generation if nothing more is generated.

Saibo-creator committed:
Merge branch 'huggingface:main' into feature/cfgcd

Saibo-creator commented:
> Hi @Saibo-creator 👋
> 
> It's great to see a project with a working example! I'd love to add it to `transformers` at some point, but we don't have the capacity to maintain a new text generation project at the moment -- you can probably see from my response time in the PRs that our bandwidth at the moment is quite limited :) Since `transformers` is used in production, we can't add features if we don't have the capacity to maintain them.
> 
> My suggestion: let's add the code as is under `/examples/research_projects/grammar`, for which the `transformers` team has 0 maintenance guarantees, and move it into the main `transformers` folder as soon as we have capacity on our end. Does it sound good to you? 🤗
> 
> P.S.: as a research project, you'd be able to make any changes you want with pretty much no barriers on our side ;)

Sounds great! Thank you @gante !

gante commented:
@Saibo-creator sounds great!

And don't let my conservative approach to your suggestions lower your enthusiasm, I'm enjoying your contributions :D

Saibo-creator committed:
Merge branch 'huggingface:main' into feature/cfgcd

Saibo-creator committed:
introduce transformers-CFG README.md

Saibo-creator committed:
add example in README.md

oobabooga commented:
I have tested the code in this PR and found that it works very nicely, so I borrowed it for my repository (with due credits): https://github.com/oobabooga/text-generation-webui/pull/4953

It is more robust than the [torch-grammar](https://github.com/Shopify/torch-grammar) EBNF implementation that I was previously using, which would half the time throw and error while importing a seemingly valid grammar.

Being able to generate structured output like json and lists for a given prompt has many practical applications and this logits processor makes that easy to setup, so I find it extremely valuable.

abhinavkulkarni commented:
Thanks @oobabooga, I was also able to test it successfully in HuggingFace TGI. It does work very well.

lukolszewski commented:
> Thanks @oobabooga, I was also able to test it successfully in HuggingFace TGI. It does work very well.

Is that code on github somewhere perhaps?

Emekaborisama commented:
i did set this up on fastapi and it only return result once

Saibo-creator commented:
Could you give a working example to show the problem ? I would be happy to investigate it.

Saibo-creator commented:
Since Transformers will not merge this in the near future, I have written a small  extension library. The use is very straightforward.

https://github.com/Saibo-creator/transformers-CFG/tree/main

mghiasvandm commented:
Hello! My use case requires the grammar to be dependent on the input text. I'm wondering if the current implementation supports passing a batch of grammars along with the batch of input and constrain the output based on different grammars ?

Saibo-creator commented:
> Hello! My use case requires the grammar to be dependent on the input text. I'm wondering if the current implementation supports passing a batch of grammars along with the batch of input and constrain the output based on different grammars ?

Hey! This is an interesting use case and I'm working on it. Will keep you updated.

nidhishgajjar commented:
> **Orb Code Review** (powered by GLM 5.1 on [Orb Cloud](https://orbcloud.dev))

Thanks for this contribution! Grammar-constrained decoding is a highly requested feature. I've reviewed the core implementation, logits processor, and tests. Here are my findings:

---

### 🚨 Must-Fix: README.md replaced with fork description

The diff replaces the entire `README.md` with just 3 lines about a "Transformers-CFG" fork. This is almost certainly unintentional and should be reverted. Only the relevant code changes should be included in this PR.

---

### 🐛 Bug: `filter_logits` doesn't return modified scores

```python
def filter_logits(self, logits, device):
    acceptance = self.grammar_constraint.batch_filter_vocab(self.batch_stacks, device)
    logits[~acceptance] = -math.inf
    # Missing: return logits
```

While in-place mutation works here, the method is called without using its return value in `process_logits`, which then returns `scores`. Since `logits` is the same object as `scores`, this works, but it's fragile — the method should either return `logits` or be clearly named to indicate mutation (e.g., `filter_logits_inplace`).

---

### ⚠️ Fragile tokenizer detection in `TokenTrie.load_tokens`

```python
if "gpt2" in tokenizer.__class__.__name__.lower():
    ...
elif "llama" in tokenizer.__class__.__name__.lower():
    ...
else:
    print("Warning: unrecognized tokenizer...")
```

This is fragile — any tokenizer with "gpt2" or "llama" in its class name would match. Consider using `isinstance` checks or a registry pattern. Also, using `print()` for warnings instead of `logger.warning()` is inconsistent with the rest of the file.

---

### ⚠️ Debugging/profiling code left in production code

```python
# In GrammarConstraint.__init__:
self.tt = 0
self.nt = 0

# In token_acceptance_for_stack:
st = time.time()
...
et = time.time() - st
self.tt += et
self.nt += 1
```

This timing/profiling code should be removed before merging, or at least gated behind a debug flag.

---

### ⚠️ `parse_ebnf` silently swallows errors

```python
except RuntimeError as err:
    logger.warning("error parsing grammar:", err)
    return ParseState()
```

Returning an empty `ParseState()` on parse errors will cause confusing downstream errors (KeyError on missing rule IDs, etc.). Consider raising the error or at minimum returning `None` so callers can detect the failure.

---

### ⚠️ `accept_token_id` raises generic `Exception`

```python
raise Exception(
    f"At least one of the stack should be empty when EOS is reached..."
)
```

Should use a specific exception type (e.g., `ValueError`).

---

### ⚠️ Test quality concerns

- Tests use plain functions instead of pytest conventions (no `@pytest.mark`, no fixtures)
- Tests download models without `@slow` markers
- `test_grammar_constrained_decoding_greedy_w_balanced_parenthesis_grammar` uses `do_sample=True` (non-deterministic) despite the name saying "greedy"
- Missing `no_newline` at end of file

---

### 📝 Minor Notes

- `StaticGrammarConstraint` is an empty stub — consider removing it or adding a `NotImplementedError` with a clear message
- `if __name__ == "__main__"` block in library code (`grammar_utils.py`) is unusual — should this be an example instead?
- The `lru_cache(maxsize=32768)` on `advance_stack` and `token_acceptance_for_stack` could grow large for complex grammars. Consider whether unbounded caching is appropriate.
- `token_acceptance_for_stack` takes `device` as a cache key parameter, which means the same computation is duplicated for different devices

---

### Summary

The EBNF parsing, stack-based grammar evaluation, and token trie approach are well-designed and the feature is valuable. However, the README must be reverted, and the debugging code, fragile tokenizer detection, and silent error handling should be cleaned up before merging.

### Assessment: request-changes

_This review was created by an AI assistant (OpenHands) on behalf of Orb Cloud._
